### DISCUSSION TITLE
file version "prodvers" values do not appear to be used in exe info

### DISCUSSION THREAD

**Discussion by dr4kk4r** [pyinstaller/pyinstaller#7336] [Help]
When editing file version info prodvers values (e.g. prodvers=(10, 11, 12, 0)) do not seems to make their way into exe info.
At this time the only way to change product version is to modify value in this line: StringStruct('ProductVersion', '4.5.6') .

Is prodvers used for a different reason that isn't obvious to me?

**Comment by rokm** [comment]
If you read the [VS_VERSIONINFO](https://learn.microsoft.com/en-us/windows/win32/menurc/vs-versioninfo) structure from the generated executable (for example, using PyInstaller's `pyi-grab_version` utility), you'll see that `prodvers` was in fact written there as part of the [VS_FIXEDFILEINFO](https://learn.microsoft.com/en-us/windows/win32/api/verrsrc/ns-verrsrc-vs_fixedfileinfo) sub-structure. 

So the question is whether the utility that you use to display/check version info reads from the fixed file-info structure or from the dynamic string structures. For example, "Details" tab in Explorer seems to display `filevers` from fixed structure as "File version" and the dynamic string field "ProductVersion" as "Product Version".

As far as I understand, `filevers` and `prodvers` are intended for direct numerical comparison, whereas dynamic string fields are intended for display in UI.

**Reply by dr4kk4r** [reply]
Thanks for a quick reply.

"For example, "Details" tab in Explorer seems to display `filevers` from fixed structure as "File version" and the dynamic string field "ProductVersion" as "Product Version".

That is what I see on my end as well. I just don't understand the inconsistency and or intent of this behavior.

What's the best way to pull out `prodvers` from and within exe? I was hoping that a build exe would report `prodvers` via some command line argument, e.g. `--version` . It looks like that has to be implemented by the developer unless I am missing something.

**Reply by rokm** [reply]
> What's the best way to pull out `prodvers` from and within exe? I was hoping that a build exe would report `prodvers` via some command line argument, e.g. `--version` . It looks like that has to be implemented by the developer unless I am missing something.

Yeah, you need to implement that functionality in your program code, since command-line parsing is largely outside the scope of PyInstaller's bootloader.

You'll have to read the VS_FIXEDFILEINFO from the executable, one way or another.

One possibility is to use the `pefile`:

```python
import pefile
import sys

pe = pefile.PE(sys.executable)
if len(pe.VS_FIXEDFILEINFO) == 1:
    fixed_info = pe.VS_FIXEDFILEINFO[0]
    product_version = (
        (fixed_info.ProductVersionMS & 0xFFFF0000) >> 16,
        (fixed_info.ProductVersionMS & 0x0000FFFF),
        (fixed_info.ProductVersionLS & 0xFFFF0000) >> 16,
        (fixed_info.ProductVersionLS & 0x0000FFFF),
    )
    print("Product version: %d.%d.%d.%d" % product_version)
```

The other is to use win32api (e.g., the [`GetFileVersionInfo`](https://learn.microsoft.com/en-us/windows/win32/api/winver/nf-winver-getfileversioninfow) function), either manually bound via `ctypes` or using `pywin32`.

**Reply by dr4kk4r** [reply]
That works well for extracting product version within an exe. Thanks!

For insertion, is it best to generate a file version (same template as `pyi-grab_version`) in CI/CD and then do a build via spec file?
Using  `pyi-set_version` post build is also an option.

Above options require manipulating text files. Simple search and replace could work. Or is there a better way?
Calling something programmatically to modify existing version file via command line arguments would be ideal.
Something like:
`pyi-modify_version version_file_path --prodver 1.2.3.4 --filevers 5.6.7.8 -- ProductName "Some new product name"`

This way a default version file would have all necessary info and only things desired could be changed. This could help with CI/CD  when creating versioned builds.

**Reply by rokm** [reply]
> For insertion, is it best to generate a file version (same template as `pyi-grab_version`) in CI/CD and then do a build via spec file? 

I think that's the best approach, yes.

> Using `pyi-set_version` post build is also an option.

It is, but it probably invalidates the PE checksum. Our building code now computes that at the end of the build process, but when `pyi-set_version` was written, that was not the case, so the utility does not re-compute the checksum after (over)writing the version info resource.

> Above options require manipulating text files. Simple search and replace could work. Or is there a better way? 

You could also set up the version info structure in the spec file, and pass it to EXE instead of the filename. You could do this either with literal initialization, or by loading it from the text file; and then override the fields of interest with values passed via environment variables (e.g., MYBUILD_PRODVERS, MYBUILD_FILEVERS, MYBUILD_PRODUCT_NAME).

> Calling something programmatically to modify existing version file via command line arguments would be ideal. Something like: `pyi-modify_version version_file_path --prodver 1.2.3.4 --filevers 5.6.7.8 -- ProductName "Some new product name"`
> 
> This way a default version file would have all necessary info and only things desired could be changed. This could help with CI/CD when creating versioned builds.

If you need that sort of (advanced) functionality, you can always write your own utility; either for generating the text file consumed by PyInstaller, or for directly overwriting the version info resource in the generated executable.

**Reply by dr4kk4r** [reply]
> If you need that sort of (advanced) functionality, you can always write your own utility; either for generating the text file consumed by PyInstaller, or for directly overwriting the version info resource in the generated executable.

I was able to overwrite the version info resource in the generated exe so that is an options. Could be a neat feature for pyinstaller.

> You could also set up the version info structure in the spec file, and pass it to EXE instead of the filename. You could do this either with literal initialization, or by loading it from the text file; and then override the fields of interest with values passed via environment variables (e.g., MYBUILD_PRODVERS, MYBUILD_FILEVERS, MYBUILD_PRODUCT_NAME).

I understand how to access environment variables in Python script, but I am not sure how to plug them into version file without running another py script. That script could potentially be executed in `*.spec` file. 

I also tried doing following in version file:

`filevers=(os.getenv("blah"), 8, 9, 7 ),`

But I get a `NameError: name 'os' is not defined` . Are you referring to something else?

**Reply by rokm** [reply]
Yeah, I was referring to loading and processing/modifying the version info within the spec file, not the version file, e.g.:

```python
# -*- mode: python ; coding: utf-8 -*-
# myprogram.spec
import sys
import os

if sys.platform == 'win32':
    from PyInstaller.utils.win32.versioninfo import load_version_info_from_text_file   # Requires #7340 or equivalent loading function in the spec file
    version_file = os.path.abspath(os.path.join(SPECPATH, 'file_version_info.txt'))
    version = load_version_info_from_text_file(version_file)
    if 'MYPROGRAM_PRODVERS' in os.environ:
        prodvers = [int(x) for x in os.environ['MYPROGRAM_PRODVERS'].split('.')]  # "1.2.3.4" -> [1,2,3,4]
        assert len(prodvers) == 4
        print(f"Overriding prodvers with {prodvers!r}")
        version.ffi.productVersionMS = prodvers[0] << 16 | prodvers[1]
        version.ffi.productVersionLS = prodvers[2] << 16 | prodvers[3]
else:
    version = None

...

a = Analysis(
...

exe = EXE(
   ...
    version=version,
)
...
```

**Reply by dr4kk4r** [reply]
Thank you for multiple clear explanations.

**Fantastic support**.

**Reply by dr4kk4r** [reply]
> from PyInstaller.utils.win32.versioninfo import load_version_info_from_text_file

For some reason I don't have that function. I am on pyinstaller `5.7.0` and Python `3.8.5` .

**Reply by rokm** [reply]
That function was added in #7340, and will be available in the next release. 

In the meantime, you need to implement its equivalent in the spec file - you can just copy implementation from #7340, but you need to add `from PyInstaller.utils.win32.versioninfo import VSVersionInfo` for deserialization via `eval` to work.

**Reply by dr4kk4r** [reply]
> from `PyInstaller.utils.win32.versioninfo import VSVersionInfo` for deserialization via `eval`

That is what I did. Thanks!
